### PR TITLE
Upgrade TypeScript to v5

### PR DESCRIPTION
**Pull request by maetl**
## Description

Upgrade TypeScript version on the project (hopefully) without requiring major changes to Rollup and Jest configuration.

The main changes are documented in individual commits (happy to squash these if preferable). Getting it to build correctly required a few minor changes to the source to take into account recent changes to enum behaviour, and dealing with several small issues relating to scope ambiguity.

Some minor warnings with peer dependencies and a bunch of linter stuff in the tests could probably be looked at further.

Need to manually test under the following scenarios:

- Checkout, install and run tests from fresh clone of the repo
- Compile from TS source and build a releasable version of the package
- Install as npm package into separate TS project and successfully import `Ink.Story` and `Ink.Compiler`
- Install as npm package into separate JS project and successfully import `Ink.Story` and `Ink.Compiler`
- Copypasta compiled JS dist files into non NPM project and confirm they work when imported via script tag

## Checklist

- [x] The new code additions passed the tests (`npm test`).
- [ ] The linter ran and found no issues (`npm run-script lint`).

### CODE REVIEW

**Comment by maetl** [comment]
Linter issues seem to be mostly existing code that is not modified as part of the TypeScript upgrade (it is likely caused by upgrade to ESLint typescript config and TS version).

For now, I will avoid pushing any commits with a bunch of code formatting changes as it will make it very difficult to read and understand this PR as a whole.

**Comment by smwhr** [comment]
Thanks for this PR. I will review it in the week-end.

Can you maybe reorder the gitlab pipeline steps so that even if lint does not pass, we can see the build and tests succeed ?

**Comment by maetl** [comment]
That’s a great idea for minimising cost of review/changes here, thanks.

I’ve changed order of workflow steps so the linter runs last and added a `.prettierrc.json` with `trailingComma: "es5"` to remove hundreds of trivial warnings that were making the linter output hard to follow.

It’s now much more clear to read the output and see what specific style linting issues are a result of the upgrade.
